Adding a sync() function to filetarget

slim-bean opened the pull request:
I still want to add a test to verify the sync function will remove files properly.

I ran into a circular imports issue when adding a config to the FileTarget, because the targets package depends on the api package for ScrapeConfig, so if I put the FileTarget config in the targets package it creates a circular import.
I think it made most sense to put the TargetConfig in the api package rather than move the ScrapeConfig to the target package.  Though this leads me to think all the other configs (positions, client) should also be moved to the api package for consistency?
I'm also open to another solution if I'm missing something obvious

tomwilkie commented:
Re: config structs, I tend to put them alongside the structs that they configure, to prevent circular imports.  The only alternative is having them all in one place, and that means everything tends to import the config package, which I don't like.

tomwilkie commented on pkg/promtail/targets/filetarget.go:
```diff
@@ -198,12 +159,104 @@ func (t *FileTarget) run() {
+func (t *FileTarget) sync() error {
+
+	// Find list of directories to add to watcher.
+	var err error
```
This `var err error` is redundant, the line below will instantiate the variable too.

tomwilkie commented on pkg/promtail/targets/filetarget.go:
```diff
@@ -198,12 +159,104 @@ func (t *FileTarget) run() {
+	}
+
+	// Get the current unique set of dirs to watch.
+	dirs := make(map[string]struct{})
```
We like to avoid make and prefer the literal style initialisation (`map[string]struct{}{}`), unless you already know the size, in which case you have to use `make`.

tomwilkie commented on pkg/promtail/targets/filetarget.go:
```diff
@@ -198,12 +159,104 @@ func (t *FileTarget) run() {
+	}
+
+	// If no files exist yet watch the directory specified in the path.
+	if matches == nil {
```
`len(matches) == 0` is safer, even though Glob specifies it returns nil.  It is less brittle.

tomwilkie commented:
Re: config structs, I tend to put them alongside the structs that they configure, to prevent circular imports.  The only alternative is having them all in one place, and that means everything tends to import the config package, which I don't like.

tomwilkie commented:
Given the complexity of the sync code, I'm coming round to the idea of removing the ionotify watcher and just relying on sync.  WDYT?

slim-bean commented:
To move the TargetConfig struct back to the targets package we would have to move the ScrapeConfig out of the api package and put it somewhere else to avoid the circular import, thoughts on what a good place for it would be?

slim-bean commented:
Re: the ionotify watcher, I was playing around yesterday with out the tailer works, and when pointed at a new file it will start at the beginning and "catch up" so to speak, so the polling nature of the sync function should work just fine.  I'll think about it a little more but at the moment I don't think the event driven ionotify watcher is gaining us anything except some addition complexity and I would support removing it.

woodsaj commented:
move the targetConfig struct to the targets package and move the Config struct to its own package, eg mv `api/{config,entry_parser}.go` to `config/`

that should resolve address the circular dependency issues

slim-bean commented:
I think I'd like to bring up removing the fsnotify code in another issue/PR and get this merged as it's already getting hard to review with the refactorings

tomwilkie commented on pkg/promtail/targets/filetarget.go:
```diff
@@ -16,7 +18,7 @@ import (
 	fsnotify "gopkg.in/fsnotify.v1"
 
 	"github.com/grafana/loki/pkg/helpers"
-	"github.com/grafana/loki/pkg/promtail/api"
```
Super nit picking: go imports tend to be order; stdlib, external package, internal package.

tomwilkie commented on pkg/promtail/targets/filetarget.go:
```diff
@@ -198,12 +171,103 @@ func (t *FileTarget) run() {
+	}
+
+	// Remove any directories which no longer need watching.
+	if len(t.watches) > 0 && len(t.watches) != len(dirs) {
```
So its possible that a directory is added & removed between a sync, meaning `len(t.watches) == len(dirs)`, which would mean this is skipped.  Remove the if completely, seems like an optimisation.

tomwilkie commented on pkg/promtail/targets/filetarget.go:
```diff
@@ -198,12 +171,103 @@ func (t *FileTarget) run() {
+
+	// Add any directories which are not already being watched.
+	for dir := range dirs {
+		if _, ok := t.watches[dir]; !ok {
```
Super nit: I generally like to unnest code as much as possible, so prefer:

```golang
for dir := range dirs {
	if _, ok := t.watches[dir]; ok {
		continue
	}
	...
}
```

tomwilkie commented on pkg/promtail/targets/filetarget.go:
```diff
@@ -198,12 +171,103 @@ func (t *FileTarget) run() {
+	// Remove any directories which no longer need watching.
+	if len(t.watches) > 0 && len(t.watches) != len(dirs) {
+		for watched := range t.watches {
+			if _, ok := dirs[watched]; !ok {
```
Ditto unnest early.

tomwilkie commented on pkg/promtail/targets/filetarget.go:
```diff
@@ -198,12 +171,103 @@ func (t *FileTarget) run() {
+
+	// Start tailing all of the matched files if not already doing so.
+	for _, p := range matches {
+		if _, ok := t.tails[p]; !ok {
```
Unnest early.

tomwilkie commented on pkg/promtail/targets/filetarget.go:
```diff
@@ -198,12 +171,103 @@ func (t *FileTarget) run() {
 	}
 }
 
+func (t *FileTarget) sync() error {
```
This function is a bit of a beast, and might benefit from breaking up.

Also, so set arithmetic helpers function (like subtraction, a - b) might help cleanup the logic:

```
dirs, files := glob()

toStartWatching := sub(t.watches, dirs)
startWatching(toStartWatching)

toStopWatching := sub(dirs, t.watches)
stopWatching(toStopWatching)

toStartTailing := sub(t.tailers, files)
startTailing(toStartTailing)

toStopTailing := sub(files, t.tailers)
stopTailing(toStopTailing)
```

Might make it more obvious and easier to follow.  Just a thought.

slim-bean commented:
Updated based on feedback, cleaned up the sync function and simplified as much as I could, also attempted to reuse code between the fsnotify watcher stuff so there shouldn't be any duplicated logic in there as well either.

tomwilkie commented on cmd/promtail/main.go:
```diff
@@ -5,6 +5,7 @@ import (
 	"os"
 
 	"github.com/go-kit/kit/log/level"
+	"github.com/grafana/loki/pkg/promtail/config"
```
This import is in the wrong place.

tomwilkie commented on pkg/promtail/config/config.go:
```diff
@@ -0,0 +1,48 @@
+	"path/filepath"
+
+	"github.com/grafana/loki/pkg/promtail/scrape"
+	"github.com/grafana/loki/pkg/promtail/targets"
```
These imports are in the wrong place.

tomwilkie commented on pkg/promtail/config/config.go:
```diff
@@ -0,0 +1,48 @@
+	"github.com/grafana/loki/pkg/promtail/scrape"
+	"github.com/grafana/loki/pkg/promtail/targets"
+	"gopkg.in/yaml.v2"
+
```
This newline is redundant.

tomwilkie commented on pkg/promtail/promtail.go:
```diff
@@ -2,9 +2,9 @@ package promtail
 
 import (
 	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/grafana/loki/pkg/promtail/config"
```
Import in wrong place.

tomwilkie commented on pkg/promtail/promtail_test.go:
```diff
@@ -0,0 +1,225 @@
+
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/cortexproject/cortex/pkg/util/flagext"
+	"github.com/go-kit/kit/log"
```
Can we have these in two groups (external & internal imports)? Separated with a newline.

tomwilkie commented on pkg/promtail/scrape/scrape.go:
```diff
@@ -0,0 +1,35 @@
+import (
+	"fmt"
+
+	"github.com/grafana/loki/pkg/promtail/api"
```
Two groups pls.

tomwilkie commented on pkg/promtail/targets/filetarget.go:
```diff
@@ -9,14 +10,14 @@ import (
 
 	"github.com/grafana/loki/pkg/helpers"
 	"github.com/grafana/loki/pkg/promtail/api"
+
```
No newline here pls.

tomwilkie commented on pkg/promtail/targets/filetarget.go:
```diff
@@ -168,164 +142,161 @@ func (t *FileTarget) run() {
-		}
-		close(t.done)
-	}()
+	t.updateTailers(matches)
```
As this is only called here, I wouldn't make it a separate func.

tomwilkie commented on pkg/promtail/targets/filetarget.go:
```diff
@@ -168,164 +142,161 @@ func (t *FileTarget) run() {
+// Returns the elements from set b which are missing from set a
+func missing(a map[string]struct{}, b map[string]struct{}) map[string]struct{} {
+	c := map[string]struct{}{}
+	for dir := range b {
```
Super nit: `dir` name is a bit weird.  I tend to go with `as`, `bs` and then call them `a` and `b` in the loop.

tomwilkie commented on pkg/promtail/targets/filetarget.go:
```diff
@@ -168,164 +142,161 @@ func (t *FileTarget) run() {
+	for dir := range b {
+		if _, ok := a[dir]; ok {
+			continue
+		} else {
```
Don't need the else with a continue.

tomwilkie commented on pkg/promtail/targets/filetarget.go:
```diff
@@ -168,164 +142,161 @@ func (t *FileTarget) run() {
+func (t *FileTarget) updateTailers(matches []string) {
+	// Start tailing all of the matched files if not already doing so.
+	for _, p := range matches {
+		err := t.startTailing(p)
```
Is it safe to start tailing things we already tail?  Oh I see it is.  Can you use the same pattern as in `sync()`?

slim-bean replied:
can you elaborate a little with what you mean about using the same pattern as sync()?

tomwilkie commented:
Sorry @slim-bean!  Super nit picky.  But its super close now.

tomwilkie commented on .gitignore:
```diff
@@ -7,3 +7,4 @@ cmd/loki/loki
 cmd/promtail/promtail
 /loki
 /promtail
+.idea/
```
Whats this?  Probably belongs in ~/.gitignore and then do `git config --global core.excludesfile $HOME/.gitignore`

tomwilkie commented on pkg/promtail/targets/filetarget_test.go:
```diff
@@ -27,8 +29,7 @@ func TestLongSyncDelayStillSavesCorrectPosition(t *testing.T) {
 	if err != nil {
-		t.Error(err)
-		return
+		t.Fatal(err)
```
Should have mentioned this earlier, but I prefer `require.NoError(t, err)` instead of these 3 lines.  We can come back and tidy this up.
